test/distributed/fsdp/test_fsdp_comm.py

```diff
@@ -283,7 +288,102 @@ def reset_mocks():
+                self.mlp2 = MLP(dim)
+                self.mlp3 = MLP(dim)
+
+            def forward(self, ys: List[torch.Tensor], works: List[dist.Work]):
+                (y1, y2, y3), (work1, work2, work3) = ys, works
+                work1.wait()
+                z1 = self.mlp1(y1)
+                work2.wait()
+                z2 = self.mlp2(y2)
+                work3.wait()
+                z3 = self.mlp3(y3)
+                return z1 + z2 + z3
+
+        class ReduceModel(nn.Module):
+            def __init__(self, dim: int, group: dist.ProcessGroup):
+                super().__init__()
+                self.reduce_module1 = ReduceModule(dim, group)
+                self.reduce_module2 = ReduceModule(dim, group)
+                self.reduce_module3 = ReduceModule(dim, group)
+                self.mlps = MLPs(dim)
+
+            def forward(self, x: torch.Tensor):
+                y1, work1 = self.reduce_module1(x)
+                if isinstance(self.mlps.mlp1, FSDP):
+                    self.mlps.mlp1._unshard(async_op=True)
+                y2, work2 = self.reduce_module2(x)
+                if isinstance(self.mlps.mlp2, FSDP):
+                    self.mlps.mlp2._unshard(async_op=True)
+                y3, work3 = self.reduce_module3(x)
+                if isinstance(self.mlps.mlp3, FSDP):
+                    self.mlps.mlp3._unshard(async_op=True)
+                return self.mlps([y1, y2, y3], [work1, work2, work3])
+
+        group = self.process_group
+        batch_size, dim = 2, 8
+        torch.manual_seed(42)
+        ref_model = DDP(ReduceModel(dim, group).cuda(), device_ids=[self.rank])
+        ref_optim = torch.optim.Adam(ref_model.parameters(), lr=1e-2)
+
+        torch.manual_seed(42)
+        model = ReduceModel(dim, group)
+        model.mlps = FSDP(
+            model.mlps,
+            sharding_strategy=ShardingStrategy.SHARD_GRAD_OP,
+            auto_wrap_policy=ModuleWrapPolicy((MLP,)),
+            device_id=self.rank,
+            use_orig_params=use_orig_params,
+        )
+        model.mlps.check_is_root()
+        mlp_params = set(model.mlps.parameters())
+        mlp_param_names = {n for n, p in model.named_parameters() if p in mlp_params}
+        DDP._set_params_and_buffers_to_ignore_for_model(model, mlp_param_names)
+        model = DDP(model.cuda(), device_ids=[self.rank])
+        optim = torch.optim.Adam(model.parameters(), lr=1e-2)
+
+        torch.manual_seed(42 + self.rank + 1)
+        inp = torch.randn((batch_size, dim), device="cuda")
+
+        for _ in range(10):
+            losses: List[torch.Tensor] = []
+            for _model, _optim in ((ref_model, ref_optim), (model, optim)):
+                losses.append(_model(inp).sum())
+                losses[-1].backward()
+                _optim.step()
+                _optim.zero_grad()
+            self.assertEqual(losses[0], losses[1])
+            model.module.mlps._wait_unshard_streams_on_current_stream()
```

Comment:
how could we tell in the test that the allreduce is overlapped with the unshard? or did you check this in trace?

Reply:
had to check in the trace manually since it is hard to check it programmatically